### PR TITLE
Toggle question text summary section

### PR DESCRIPTION
**Pull request by JamesGardiner**
### What is the context of this PR?
Describe what you have changed and why, link to other PRs or Issues as appropriate.

Added a 'hide_on_section_summary' key to question objects in schema. When set to true this block will not be included in rendered summary sections.

### How to review 
Complete the Who Lives Here census sections. Assert that only list collectors are displayed on the seciton summary.

### CODE REVIEW

File: app/views/contexts/summary/group.py [hunk 1]
```diff
@@ -28,7 +28,11 @@ def _build_blocks(
         block_ids_on_path = [location.block_id for location in path]
 
         for block in group_schema['blocks']:
-            if block['id'] in block_ids_on_path and block['type'] == 'Question':
+            if (
+                block['id'] in block_ids_on_path
+                and block['type'] == 'Question'
+                and not block.get('question', {}).get('hide_on_section_summary', False)
```
Comment (MebinAbraham) [inline app/views/contexts/summary/group.py]: Does `hide_on_section_summary` need to live within `question`?. We won't have more than 1 question per block hence we can define it at the block level?. And this currently wouldn't work for `question_variants`?

File: data-source/jsonnet/england-wales/household/blocks/who-lives-here/any_visitors.jsonnet [hunk 2]
```diff
@@ -18,6 +18,7 @@ function(census_date) {
     id: 'any-visitors-question',
     title: questionTitle(census_date),
     mandatory: true,
+    hide_on_section_summary: true,
```
Comment (MebinAbraham) [inline data-source/jsonnet/england-wales/household/blocks/who-lives-here/any_visitors.jsonnet]: `Does anyone usually live at` question need to also be hidden from the summary. Also these need to be hidden on all census schema, ie ENG/WLS/NIR.

Screenshot 2019-09-03 at 09 12 11

**Review by MebinAbraham — CHANGES_REQUESTED** [review]
The functionality works as expected.

Just some minor comments.
Should we extend an existing functional test for this?

File: app/views/contexts/summary/group.py (hunk 1 above)
Comment (JamesGardiner) [reply app/views/contexts/summary/group.py]: Are you suggesting have this as an allowable key only on Question type blocks?

Comment (MebinAbraham) [reply app/views/contexts/summary/group.py]: I think that would make sense?. Defined at the block level and validator can ensure it is defined on allowed types.

Comment (JamesGardiner) [reply app/views/contexts/summary/group.py]: Yeah ok, it's a small change, and would extend the functionality to question_variants

File: scripts/test_schemas.sh [hunk 3]
```diff
@@ -6,7 +6,7 @@ if [ "$1" == "--local" ] || [ "$2" == "--local" ]; then
 fi
 
 if [ "$run_docker" == true ]; then
-    branch=v3
+    branch=add-hide-on-section-summary-to-questions
```
Comment (JamesGardiner) [inline scripts/test_schemas.sh:9]: Revert before merge

File: tests/app/views/contexts/summary/test_group.py [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+
+
+def test_group_serialize_hide_on_summary_false(
+    app, mocker, schema, path
+):  # pylint: disable=unused-argument
+    answer_store = mocker.MagicMock()
+    list_store = mocker.MagicMock()
+    metadata = mocker.MagicMock()
+
+    schema['blocks'][0]['hide_on_section_summary'] = False
```
Comment (LJBabbage) [inline tests/app/views/contexts/summary/test_group.py]: As discussed, as we would never see in a normal schema "hide_on_section_summary": False, not sure we should do this in a test

File: data-source/jsonnet/england-wales/household/blocks/who-lives-here/anyone_usually_live_at.jsonnet [hunk 5]
```diff
@@ -23,6 +23,7 @@ function(census_date) {
       when: [rules.hasPrimary],
     },
   ],
+  hide_on_section_summary: true,
```
Comment (LJBabbage) [inline data-source/jsonnet/england-wales/household/blocks/who-lives-here/anyone_usually_live_at.jsonnet]: Any reason we mix case, in the schema its capitalised or is this a jsonnet thing?

File: app/views/contexts/summary/group.py [hunk 6]
```diff
@@ -28,7 +28,11 @@ def _build_blocks(
         block_ids_on_path = [location.block_id for location in path]
 
         for block in group_schema['blocks']:
-            if block['id'] in block_ids_on_path and block['type'] == 'Question':
+            if (
+                block['id'] in block_ids_on_path
+                and block['type'] == 'Question'
+                and not block.get('hide_on_section_summary', False)
```
Comment (LJBabbage) [inline app/views/contexts/summary/group.py]: Enjoyable, 99% of the time this will default and then satisfy the condition :smiley:

**Review by LJBabbage — CHANGES_REQUESTED** [review]

**Comment by JamesGardiner** [comment]
[Remove summary of the questions from WLH summary page at the top (M)](https://trello.com/c/LPmBf6D9/3257-remove-summary-of-the-questions-from-wlh-summary-page-at-the-top-m)

File: tests/app/views/contexts/summary/test_group.py (hunk 4 above)
Comment (JamesGardiner) [reply tests/app/views/contexts/summary/test_group.py]: That's not what I thought we agreed.

There are three scenarios possible:
  - "hide_on_section_summary": true
  - "hide_on_section_summary": false
  - Hide on section summary omitted defaulting to false.

I'll add a further test that covers the third scenario in addition to the above.

File: data-source/jsonnet/england-wales/household/blocks/who-lives-here/anyone_usually_live_at.jsonnet (hunk 5 above)
Comment (JamesGardiner) [reply data-source/jsonnet/england-wales/household/blocks/who-lives-here/anyone_usually_live_at.jsonnet]: Can you be more specific? Do you mean the case of the boolean value?

File: tests/app/views/contexts/summary/test_group.py (hunk 4 above)
Comment (JamesGardiner) [reply tests/app/views/contexts/summary/test_group.py]: Scratch that. I've already tested all three scenarios... :man_facepalming:

File: data-source/jsonnet/england-wales/household/blocks/who-lives-here/anyone_usually_live_at.jsonnet (hunk 5 above)
Comment (LJBabbage) [reply data-source/jsonnet/england-wales/household/blocks/who-lives-here/anyone_usually_live_at.jsonnet]: Yep

Comment (JamesGardiner) [reply data-source/jsonnet/england-wales/household/blocks/who-lives-here/anyone_usually_live_at.jsonnet]: JSON standard is booleans as `true` or `false` - same as JS :+1:

**Review by rmccar — APPROVED** [review]

File: Pipfile [hunk 7]
```diff
@@ -30,6 +30,7 @@ fakeredis = "*"
 mypy = "*"
 black = "==19.3b0"
 pytest-flask = "*"
+pytest-mock = "*"
```
Comment (iwootten) [inline Pipfile]: Has this module been introduced as part of tests that have now been removed? 

Looks like its for wrapping mock.patch, which isn't used in any of these tests. I can switch to a standard mock and the tests still pass.

File: data-source/jsonnet/england-wales/household/blocks/who-lives-here/any_visitors.jsonnet [hunk 8]
```diff
@@ -13,6 +13,7 @@ local questionTitle(census_date) = {
 function(census_date) {
   type: 'Question',
   id: 'any-visitors',
+  show_on_section_summary: false,
```
Comment (iwootten) [inline data-source/jsonnet/england-wales/household/blocks/who-lives-here/any_visitors.jsonnet:16]: I know this was discussed in the office, but didn't mention it as I felt it should be part of another PR. Has anyone actually requested we should flip this flag or what it should be named?

**Review by iwootten — CHANGES_REQUESTED** [review]

File: data-source/jsonnet/england-wales/household/blocks/who-lives-here/any_visitors.jsonnet (hunk 8 above)
Comment (JamesGardiner) [reply data-source/jsonnet/england-wales/household/blocks/who-lives-here/any_visitors.jsonnet:16]: I spoke to @ajmaddaford about it and agreed to flip it

File: Pipfile (hunk 7 above)
Comment (JamesGardiner) [reply Pipfile]: Did you uninstall it and run the tests?

Comment (iwootten) [reply Pipfile]: Yes.

Comment (JamesGardiner) [reply Pipfile]: Looks like this is a result of using pipenv.

To see the actual dependency on this try:
```
$ pipenv --rm
$ pipenv install --dev
$ make test-unit
```

Comment (JamesGardiner) [reply Pipfile]: Removing the package from the pipfile and updating doesn't remove it from site-packages, so it's still available to pytest

**Review by LJBabbage — APPROVED** [review]
All good, tried it on a couple of other schemas and it worked as expected

File: Pipfile (hunk 7 above)
Comment (iwootten) [reply Pipfile]: This still works for me when I've removed the module from the pipfile, lock, completely removed/reinstalled the env and replaced mocker with mock.

What behaviour do the new unit tests use from this module?

Comment (JamesGardiner) [reply Pipfile]: Using the mocker fixture

Comment (JamesGardiner) [reply Pipfile]: Using the steps you've outlined results in a failure for me locally

File: tests/app/views/contexts/summary/test_group.py [hunk 9]
```diff
@@ -0,0 +1,132 @@
...
+        Location('section', 'number-question-one'),
+        Location('section', 'number-question-two'),
+    ]
+
+
+def test_group_serialize_hide_on_summary_true(
+    app, mocker, schema, path
+):  # pylint: disable=unused-argument
+
+    answer_store = mocker.MagicMock()
```
Comment (MebinAbraham) [inline tests/app/views/contexts/summary/test_group.py]: Can `MagicMock` from `mock` not be used?

**Review by MebinAbraham — APPROVED** [review]

File: tests/app/views/contexts/summary/test_group.py (hunk 9 above)
Comment (JamesGardiner) [reply tests/app/views/contexts/summary/test_group.py]: Yes, as the MagicMock seen here _is_ a MagicMock from stdlib. However, if we have a long term goal of moving away from unittest to pytest, we should embrace where possible and sensible pytest plugins that can give cleaner test code. See https://github.com/pytest-dev/pytest-mock/ for examples of where the mocker fixture may help, especially when patching.

Comment (iwootten) [reply tests/app/views/contexts/summary/test_group.py]: Does this module actually make the code in the PR any cleaner over and above mock?

I don't think we should be introducing dependencies if we're not actually using the features they provide.

Comment (JamesGardiner) [reply tests/app/views/contexts/summary/test_group.py]: Having the mock available as a common fixture across unit tests _is_ using part of the functionality they provide.

File: Pipfile (hunk 7 above)
Comment (iwootten) [reply Pipfile]: You don't need to use mock at all, these can be empty List/AnswerStore/dicts?

File: tests/app/views/contexts/summary/test_group.py (hunk 9 above)
Comment (ajmaddaford) [reply tests/app/views/contexts/summary/test_group.py]: I think if we want to change the approach to mocks, we should do it across all tests in a separate pull request. Otherwise it's confusing when looking at the codebase why it's done differently.

**Review by MebinAbraham — CHANGES_REQUESTED** [review]
I don't think we should have both `hide_on_section_summary` and `show_on_section_summary`. We should choose one?

**Comment by JamesGardiner** [comment]
@MebinAbraham I was under the impression `show_on_section_summary` had been chosen?

**Comment by MebinAbraham** [comment]
> @MebinAbraham I was under the impression `show_on_section_summary` had been chosen?

That's fine, but there are still references to `hide_on_section_summary` which makes things confusing.

File: tests/app/views/contexts/summary/test_group.py (hunk 9 above)
Comment (JamesGardiner) [reply tests/app/views/contexts/summary/test_group.py]: Done :+1:

**Comment by JamesGardiner** [comment]
@MebinAbraham - this should be changed in https://github.com/ONSdigital/eq-survey-runner/pull/2348#pullrequestreview-285384242

File: data-source/jsonnet/northern-ireland/household/blocks/who-lives-here/visitor_list_collector.jsonnet [hunk 10]
```diff
@@ -190,5 +190,5 @@ function(census_date) {
     add_link_text: 'Add a visitor',
     empty_list_text: 'There are no visitors',
   },
-  hide_on_section_summary: false,
+  show_on_section_summary: false,
```
Comment (MebinAbraham) [inline data-source/jsonnet/northern-ireland/household/blocks/who-lives-here/visitor_list_collector.jsonnet]: Should this not be `true`?

File: app/questionnaire/questionnaire_schema.py [hunk 11]
```diff
@@ -163,8 +163,8 @@ def get_visible_list_blocks_for_section(section):
         list_collector_blocks = []
         for group in section['groups']:
             for block in group['blocks']:
-                hidden = block.get('hide_on_section_summary', False)
-                if block['type'] == 'ListCollector' and not hidden:
+                shown = block.get('show_on_section_summary', False)
```
Comment (iwootten) [inline app/questionnaire/questionnaire_schema.py]: The default should be flipped as well.

**Review by MebinAbraham — APPROVED** [review]

**Review by iwootten — APPROVED** [review]
